src/lib/prov/bearssl/bearssl_ec.cpp

```diff
@@ -0,0 +1,194 @@
+         if (m_hash == NULL)
+            throw Lookup_Error("BearSSL ECDSA does not support EMSA " + emsa);
+
+         const SCAN_Name req(emsa);
+         m_hf = make_bearssl_hash(req.arg(0));
+         if (m_hf == NULL)
+            throw Lookup_Error("BearSSL ECDSA does not support hash " + req.arg(0));
+
+         const secure_vector<uint8_t> enc = EC2OSP(ecdsa.public_point(), PointGFp::UNCOMPRESSED);
+         m_key.qlen = enc.size();
+         m_key.q = new uint8_t[m_key.qlen];
+         memcpy(m_key.q, (unsigned char *)enc.data(), m_key.qlen);
+         m_key.curve = curve;
+         }
+
+      void update(const uint8_t msg[], size_t msg_len) override
+         {
+         m_hf->update(msg, msg_len);
+         }
+
+      bool is_valid_signature(const uint8_t sig[], size_t sig_len) override
+         {
+         const size_t order_bytes = (m_order_bits + 7) / 8;
+         if (sig_len != 2 * order_bytes)
+            return false;
+         secure_vector<uint8_t> msg = m_hf->final();
+
+         br_ecdsa_vrfy engine = br_ecdsa_vrfy_raw_get_default();
+         if (!engine(&br_ec_prime_i31, msg.data(), msg.size(), &m_key, sig, sig_len))
+             return false;
+
+         return true;
+         }
+
+      size_t max_input_bits() const { return m_order_bits; }
+
+   private:
+      br_ec_public_key m_key;
+      std::unique_ptr<HashFunction> m_hf;
+      const br_hash_class *m_hash;
+      size_t m_order_bits;
+   };
+
+class BearSSL_ECDSA_Signing_Operation : public PK_Ops::Signature
+   {
+   public:
+      BearSSL_ECDSA_Signing_Operation(const ECDSA_PrivateKey& ecdsa, const std::string& emsa) :
+         m_order_bits(ecdsa.domain().get_order().bits())
+         {
+         const int curve = BearSSL_EC_curve_for(ecdsa.domain());
+         if(curve < 0)
+            throw Lookup_Error("BearSSL ECDSA does not support this curve");
+
+         m_hash = BearSSL_hash_class_for(emsa);
+         if (m_hash == NULL)
+            throw Lookup_Error("BearSSL ECDSA does not support EMSA " + emsa);
+
+         const SCAN_Name req(emsa);
+         m_hf = make_bearssl_hash(req.arg(0));
+         if (m_hf == NULL)
+            throw Lookup_Error("BearSSL ECDSA does not support hash " + req.arg(0));
+
+         m_key.xlen = ecdsa.private_value().bytes();
+         m_key.x = new uint8_t[m_key.xlen];
+         ecdsa.private_value().binary_encode(m_key.x);
+         m_key.curve = curve;
+         }
+
+
+      void update(const uint8_t msg[], size_t msg_len) override
+         {
+         m_hf->update(msg, msg_len);
+         }
+
+      secure_vector<uint8_t> sign(RandomNumberGenerator&) override
+         {
+         const size_t order_bytes = (m_order_bits + 7) / 8;
+         secure_vector<uint8_t> sigval(2*order_bytes);
+         size_t sign_len;
+
+         br_ecdsa_sign engine = br_ecdsa_sign_raw_get_default();
+         sign_len = engine(&br_ec_prime_i31, m_hash, m_hf->final().data(), &m_key, sigval.data());
```

Comment:
You need `sign_len` here first, so it's better to declare and define it here instead of declaring it above an leaving it unitialized. That may not be of a problem now, but may become when you refactor code in the future.